File: 0011-container-with-most-water/0011-container-with-most-water.py
'''
Algorithm: Two Pointer
N = len(height)
T: O(N)
S: O(1)
'''

class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        globalMax = 0
        
        while l < r:
            localMax = r - l
            localMax *= min(height[l], height[r])
            globalMax = max(localMax, globalMax)
            
            if height[l] == height[r]:
                l += 1
                r -= 1
            elif height[l] > height[r]:
                r -= 1
            else:
                l += 1
                
        return globalMax
            
        

File: 0011-container-with-most-water/0011-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        /*
         * Algorithms & DS: #Two Pointer
         * Time: O(N), Extra Space: O(1)
         */
        int l = 0, r = height.length - 1;
        int maxArea = 0;

        while (l < r) {
            int area = (r - l) * Math.min(height[l], height[r]);

            if(height[l] < height[r]) l++;
            else r--;

            maxArea = Math.max(maxArea, area);
        }


        return maxArea;
    }
}